src/sentry/workflow_engine/processors/delayed_workflow.py

```diff
@@ -542,47 +563,67 @@ def get_groups_to_fire(
         group_id = event_key.group_id
         if event_key.workflow_id not in workflows_to_envs:
             # The workflow is deleted, so we can skip it
             continue
 
         workflow_env = workflows_to_envs[event_key.workflow_id]
+        when_result = TriggerResult.TRUE
         if when_dcg_id := event_key.when_dcg_id:
-            if not (
-                dcg := data_condition_group_mapping.get(when_dcg_id)
-            ) or not _evaluate_group_result_for_dcg(
-                dcg,
+            when_dcg = data_condition_group_mapping.get(when_dcg_id)
+            if not when_dcg:
+                continue
+            when_result = _evaluate_group_result_for_dcg(
+                when_dcg,
                 dcg_to_slow_conditions,
                 group_id,
                 workflow_env,
                 condition_group_results,
-            ):
+            )
+            if not when_result.triggered:
+                # If we're not triggering, all action-y if conditions need to be treated
+                # as tainted or not based on the when condition result.
+                if_cond_count = len(event_key.if_dcg_ids) + len(event_key.passing_dcg_ids)
+                if when_result.is_tainted():
+                    tainted += if_cond_count
+                else:
+                    untainted += if_cond_count
                 continue
 
         # the WHEN condition passed / was not evaluated, so we can now check the IF conditions
         for if_dcg_id in event_key.if_dcg_ids:
-            if (
-                dcg := data_condition_group_mapping.get(if_dcg_id)
-            ) and _evaluate_group_result_for_dcg(
-                dcg,
-                dcg_to_slow_conditions,
-                group_id,
-                workflow_env,
-                condition_group_results,
-            ):
-                groups_to_fire[group_id].add(dcg)
+            if dcg := data_condition_group_mapping.get(if_dcg_id):
+                if_result = when_result & _evaluate_group_result_for_dcg(
+                    dcg,
+                    dcg_to_slow_conditions,
+                    group_id,
+                    workflow_env,
+                    condition_group_results,
+                )
+                if if_result.is_tainted():
+                    tainted += 1
+                else:
+                    untainted += 1
+                if if_result.triggered:
+                    groups_to_fire[group_id].add(dcg)
 
         for if_dcg_id in event_key.passing_dcg_ids:
             if dcg := data_condition_group_mapping.get(if_dcg_id):
+                # TODO: Propagate taint with passing conditions.
+                if when_result.is_tainted():
+                    tainted += 1
+                else:
+                    untainted += 1
                 groups_to_fire[group_id].add(dcg)
 
-    return groups_to_fire
+    return groups_to_fire, _ConditionEvaluationStats(tainted=tainted, untainted=untainted)
```

Comment:
Nit: Wondering if we should just report the stats in here instead of returning them